src/native_ext/memory_profiling.cpp

```diff
@@ -0,0 +1,219 @@
+      Nan::Get(options, Nan::New("sampleIntervalBytes").ToLocalChecked());
+    if (
+      !maybeSampleIntervalBytes.IsEmpty() &&
+      maybeSampleIntervalBytes.ToLocalChecked()->IsNumber()) {
+      sampleIntervalBytes = Nan::To<int64_t>(maybeSampleIntervalBytes.ToLocalChecked()).FromJust();
+    }
+
+    auto maybeMaxStackDepth = Nan::Get(options, Nan::New("maxStackDepth").ToLocalChecked());
+    if (!maybeMaxStackDepth.IsEmpty() && maybeMaxStackDepth.ToLocalChecked()->IsNumber()) {
+      maxStackDepth = Nan::To<int32_t>(maybeMaxStackDepth.ToLocalChecked()).FromJust();
+    }
+  }
+
+  profiling->isRunning = profiler->StartSamplingHeapProfiler(sampleIntervalBytes, maxStackDepth);
+}
+
+NAN_METHOD(CollectHeapProfile) {
+  info.GetReturnValue().SetNull();
+
+  if (!profiling) {
+    return;
+  }
+
+  if (!profiling->isRunning) {
+    return;
+  }
+
+  v8::HeapProfiler* profiler = info.GetIsolate()->GetHeapProfiler();
+
+  if (!profiler) {
+    return;
+  }
+
+  v8::AllocationProfile* profile = profiler->GetAllocationProfile();
+
+  if (!profile) {
+    return;
+  }
+
+  auto jsResult = Nan::New<v8::Object>();
+  auto jsSamples = Nan::New<v8::Array>();
+  auto jsNodeTree = Nan::New<v8::Object>();
+  int32_t jsSamplesLength = 0;
+
+  v8::AllocationProfile::Node* root = profile->GetRootNode();
+
+  const std::vector<v8::AllocationProfile::Sample>& samples = profile->GetSamples();
+
+  profiling->generation++;
+  uint64_t generation = profiling->generation;
+
+  khash_t(SampleId)* tracking = profiling->tracking;
+
+  for (const auto& sample : samples) {
+    if (kh_get(SampleId, tracking, sample.sample_id) == kh_end(tracking)) {
+      auto jsSample = Nan::New<v8::Object>();
+      Nan::Set(
+        jsSample, Nan::New<v8::String>("nodeId").ToLocalChecked(),
+        Nan::New<v8::Uint32>(sample.node_id));
+      Nan::Set(
+        jsSample, Nan::New<v8::String>("size").ToLocalChecked(),
+        Nan::New<v8::Uint32>(uint32_t(sample.size * sample.count)));
+      Nan::Set(jsSamples, jsSamplesLength++, jsSample);
+    }
+
+    int ret;
+    khiter_t it = kh_put(SampleId, tracking, sample.sample_id, &ret);
+    if (ret != -1) {
+      kh_value(tracking, it) = generation;
+    }
+  }
+
+  for (khiter_t it = kh_begin(tracking); it != kh_end(tracking); ++it) {
+    if (!kh_exist(tracking, it)) {
+      continue;
+    }
+
+    if (kh_val(tracking, it) != generation) {
+      kh_del(SampleId, tracking, it);
+    }
+  }
+
+  StringStash stash;
+  stash.strings[V8String_Name] = Nan::New<v8::String>("name").ToLocalChecked();
+  stash.strings[V8String_ScriptName] = Nan::New<v8::String>("scriptName").ToLocalChecked();
+  stash.strings[V8String_LineNumber] = Nan::New<v8::String>("lineNumber").ToLocalChecked();
+  stash.strings[V8String_ParentId] = Nan::New<v8::String>("parentId").ToLocalChecked();
+
+  std::vector<BFSNode>& stack = profiling->stack;
+  stack.clear();
+
+  // Cut off the root node
+  for (size_t i = 0; i < root->children.size(); i++) {
```

Comment:
Could we use range-based for loop here and below as well?

Reply:
No difference really, but added it

Reply:
That's why I was asking - wondering whether there is a difference and, if not, whether we'd like to do the same thing the same way.